rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+- Fix [imprecision in string representation of floating point numbers](https://github.com/NixOS/nix/pull/6238)
+- Make the `@`-pattern consistent with destructuring
+- A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`
+- Use `,` to delimit elements of a list expression, like `[ 1, 2, 3 ]`
+- Do something about `?` meaning two different things depending on where it occurs (`{ x ? "" }:` vs `x ? ""`)
+- Better support for static analysis
+- [Syntax for hexadecimal numbers](https://github.com/NixOS/nix/pull/7695)
+
+Other discussions around language changes:
+
+- [Make path semantics less surprsing](https://github.com/NixOS/nix/issues/7338)
+- [RFC 110](https://github.com/NixOS/rfcs/pull/110) proposing a alternatives to the `with` expression
+- [Nix 2 – a hypothetical syntax](https://md.darmstadt.ccc.de/s/nix2)
+- [Nix language changes](https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d)
+
+## Drawbacks
+
+Allowing multiple language versions to coexist will over time introduce a proliferation of syntax highlighters and other tooling.
+Once the language version is accessible though, tooling can at least be adapted in a systematic way.
+
+## Alternatives
+
+- Keep the language as implemented by Nix compatible, but socially restrict the usage of undesirable features.
+    - (+) Roughly matches the current practice, no technical change needed
+    - (+) Maintains usability of old Nixpkgs versions (up to availability of fixed-output artifacts)
+    - (+) Does not break third-party codebases before making a decision, keeping Nix a dependable upstream
+        - (-) This proposal does not allow for breakages unless there is some eventual phase-out of support
+    - (-) Strict enforcement requires extra tooling that this proposal would obviate
+    - (-) The implementation of the features that are no longer desirable still incur complexity and maintenance cost
+        - (-) It's still not really possible to make changes to the language
+
+- Introduce changes to the language with language extensions or feature flags
+  - (-) Combinatorial explosion
+      - See [Haskell language extensions] for real-world experience
+  - (-) Even more maintenance overhead
+  - (+) Allows gradual adoption of features
+      - (-) We already have experimental feature flags as an orthogonal mechanism, with the added benefit that they don't incur support costs and can be dropped without loss
+
+- Never make breaking changes to the language
+    - (+) No additional maintenance effort required
+    - (-) Blocks improvements
+    - (-) Requires additions to be made very carefully
+        - (-) Even incremental changes are really expensive that way
+    - (-) Makes solving some well-known problems impossible
+
+- Continue current practice
+    - (-) There is no process for breaking changes
+    - (-) Breaking changes are not always announced
+    - (-) There are no means of determining compatibility between expressions and evaluator versions
```

Comment:
That is wrong, we do have `builtins.langVersion` as well as the ability to check for the existence of builtins. Changes to language parsing have (as far as I'm aware) been practically non-breaking so far. So Nix expressions can (largely) be written in a forwards compatible and even backwards compatible way.

The problem is rather how we are using that tool at the moment:

* Breaking changes to language have on occasion not been treated as such, i.e. implemented without incrementing `builtins.langVersion`. You can still check `builtins.nixVersion` in such cases, but it could cause us to have an user agent mess in case of competing Nix implementations. Neglecting to increase the `langVersion` number indicates that this has not been a formal part of the review process so far and possibly that the coverage of the Nix language test suite. The good news is that those are not fundamental, just operational problems.
* `builtins.langVersion` is not properly documented anywhere, but in NixOS/nix's commit messages. Lack of documentation is a known problem that is actively worked on as of late.
* Our understanding of the Nix language's behavior is still lacking, since some of it seems to be accidental, but still necessary. The problem with this is of course, that it becomes harder to properly judge if changes are breaking or not. This is improving at the moment, but we'd still need to consolidate this knowledge, so it can be used as a basis of judgement.

Reply:
> Changes to language parsing have (as far as I'm aware) been practically non-breaking so far. So Nix expressions can (largely) be written in a forwards compatible and even backwards compatible way.

That is exactly why this proposal is made. Interesting incremental changes are currently impossible because that would break everything. The proposal allows for those changes, though at substantial cost.